Sphinx gallery for example gallery

emmanuelle opened the pull request:
This PR ports our gallery examples to sphinx-gallery (http://sphinx-gallery.readthedocs.io/en/latest/).
This PR addresses #1973.

Several modifications were required to use sphinx-gallery, such as 
- using .rst instead of .txt for all source files (including generated api)
- modifiying blocks of comments in the example files 

I also had to include the source code of sphinx-gallery, since I needed to do a slight modification to one of these files. I opened a PR on sphinx-gallery (https://github.com/sphinx-gallery/sphinx-gallery/pull/119), if it gets merged in their next release we'll remove our local copy.

emmanuelle commented:
I haven't been able to use yet all the cool features of sphinx-gallery. One of them is the possibility to link automatically to examples in the API reference, see http://sphinx-gallery.readthedocs.io/en/latest/advanced_configuration.html#references-to-examples

All projects using this feature of sphinx-gallery (such as nilearn, mne-python) seem to rely on templates for the generation of their API. In scikit-image, we use ``apigen.py``, but not templates. Before hacking apigen.py, I'd like to get feedback from sphinx experts: what are the pros and cons of using either apigen.py or templates?

emmanuelle commented:
@scikit-image/core  I know that this PR is breaking Travis, but I would still welcome comments, in particular about the API generation (see my previous comment). Thanks!

emmanuelle commented:
Regarding Travis' cause of failure, I've opened an issue (https://github.com/sphinx-gallery/sphinx-gallery/issues/120) on ``sphinx-gallery``, since the error is weird: the image file is mentioned as "not readable", but is correctly included in the html page... 

Any idea about this weird error?

stefanv commented:
This is a great step forward, thanks Emmanuelle!  Can you explain what you
mean by "template"?  At the moment, apigen.py just imports scikit-image and
traverses the submodule tree, figuring which functions to autodoc, so that
we don't have to keep an index up to date when new functionality is added.

On Wed, 18 May 2016 at 22:41 Emmanuelle Gouillart <notifications@github.com>
wrote:

> Regarding Travis' cause of failure, I've opened an issue (
> sphinx-gallery/sphinx-gallery#120
> <https://github.com/sphinx-gallery/sphinx-gallery/issues/120>) on
> sphinx-gallery, since the error is weird: the image file is mentioned as
> "not readable", but is correctly included in the html page...
>
> Any idea about this weird error?
>
> —
> You are receiving this because you are on a team that was mentioned.
>
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-image/scikit-image/pull/2078#issuecomment-220232515>
>

emmanuelle commented:
Yeah, I'm super excited about it :-). Having the little pop-up when hovering over thumbnails, or links to the API in the examples, is really cool. I just have to solve this sphinx warning/error now :-/.

About templates, what I meant is that (from what I understand), the content of API is generated by ``apigen.ApiDocWriter.generate_api_doc`` (am I wrong maybe?) when ``make`` calls ``python tools/build_modref_templates.py``. ``generate_api_doc`` builds a rst string for the objects of a given module.

On the other hand, other packages (sphinx-gallery, nilearn) use templates as in  https://raw.githubusercontent.com/nilearn/nilearn/master/doc/templates/function.rst 
In particular, ``.. include:: {{module}}.{{objname}}.examples`` is used to include a small gallery of thumbnails of examples using a given function.

emmanuelle commented:
Travis is happy now, so I would be happy to get reviews :-).

jni commented on doc/examples/color_exposure/plot_adapt_rgb.py:
```diff
@@ -35,11 +35,10 @@ def sobel_each(image):
+######################################################################
+# We can use these functions as we would normally use them, but now they work
+# with both gray-scale and color images. Let's plot the results with a color
+# image:
```
@emmanuelle sorry, I haven't been following this PR, but is this format change necessary? I think it's way uglier to read!

emmanuelle replied:
@jni it is necessary indeed (strings within triple quotes are not transformed to text blocks but appear in the code). The rationale between this is that it's better to teach users to put comments with # than within triple quotes (see the discussion in https://github.com/sphinx-gallery/sphinx-gallery/issues/120), since triple quotes should be docstrings.

But it's a bit ugly to read indeed, and it was a pain to implement :-).

jni commented on doc/examples/color_exposure/plot_ihc_color_separation.py:
```diff
@@ -48,11 +48,9 @@
-"""
-.. image:: PLOT2RST.current_figure
+######################################################################
+# Now we can easily manipulate the hematoxylin and DAB "channels":
```
Getting rid of the `.. image::` directive is great, though. =)

jni commented on doc/examples/xx_applications/plot_morphology.py:
```diff
@@ -12,12 +12,19 @@
 6. Black Tophat
+
 7. Skeletonize
+
```
Another change that is rather annoying in plaintext... Is it necessary?

emmanuelle replied:
it is not, thanks (it was me puzzled about other errors and trying things to solve them...)

jni commented on doc/ext/sphinx_gallery/_static/gallery.css:
```diff
@@ -0,0 +1,127 @@
+/*
+Sphinx-Gallery is has compatible CSS to fix default sphinx themes
```
typo: "Sphinx-Gallery is has compatible CSS..."

jni commented on doc/ext/sphinx_gallery/_static/gallery.css:
```diff
@@ -0,0 +1,127 @@
+Sphinx-Gallery is has compatible CSS to fix default sphinx themes
+Tested for Sphinx 1.3.1 for all themes: default, alabaster, sphinxdoc,
+scrolls, agogo, traditional, nature, haiku, pyramid
+Tested for Read the docs theme 0.1.7 */
```
Read the docs -> Read the Docs

jni commented on doc/ext/sphinx_gallery/docs_resolv.py:
```diff
@@ -0,0 +1,435 @@
+# -*- coding: utf-8 -*-
+# Author: Óscar Nájera
```
@emmanuelle is there a reference to the original source code somewhere in this PR? It might be useful in the future in case of bug fixes/enhancements...?

emmanuelle replied:
Good idea. What would the best way to include such a reference? Add a README.txt file in the ``sphinx_gallery`` directory?

sciunto replied:
+1 for the README. However, it will be removed after the next sphinx release, right?

emmanuelle replied:
yes, after the next release of ``sphinx-gallery`` (not ``sphinx``).

jni commented on doc/ext/sphinx_gallery/docs_resolv.py:
```diff
@@ -0,0 +1,435 @@
+                  "Internet access.\nPlease check your network connection.\n"
+                  "Unable to continue embedding `{0}` links due to a URL "
+                  "Error:\n".format(this_module))
+            print(e.args)
```
Do we want print here or a real warning?

jni replied:
Ignore me. Since this is "run once" code, I don't think it matters.

sciunto commented on doc/ext/sphinx_gallery/notebook.py:
```diff
@@ -0,0 +1,123 @@
+
+
+class Notebook(object):
+    """Ipython notebook object
```
Ipython or Jupyter?

sciunto commented on doc/ext/sphinx_gallery/notebook.py:
```diff
@@ -0,0 +1,123 @@
+Parser for Jupyter notebooks
+============================
+
+Class that holds the Ipython notebook information
```
Ipython or Jupyter

JDWarner replied:
IPython, you mean ;)

emmanuelle commented:
@sciunto @jni all the files in ``ext/sphinx-gallery`` are directly copied from the ``sphinx_gallery`` package. Any changes that we make should be submitted as pull requests to the original package.

emmanuelle commented:
That is, I'm willing to submit these enhancements to sphinx-gallery but maybe this should not block this PR. 

Also, if somebody could check that it works on his/her box, that'd be great (to control that the generated gallery looks ok). I'm happy with the result, but another opinion would be good.

sciunto commented:
@emmanuelle I'm discovering your work. I have a problem to compile the doc.



```
Exception occurred:
  File "/home/fr/.local/share/virtualenvs/sk/lib/python3.5/codecs.py", line 895, in open
    file = builtins.open(filename, mode, buffering)
FileNotFoundError: [Errno 2] No such file or directory: '/home/fr/github/scikit-image/doc/source/auto_examples/transform/plot_radon_transform.txt.rst'
The full traceback has been saved in /tmp/sphinx-err-iz305edn.log, if you want to report the issue to the developers.

```

the log file

```
# Sphinx version: 1.4.1
# Python version: 3.5.1 (CPython)
# Docutils version: 0.12 release
# Jinja2 version: 2.8
# Last messages:

# Loaded extensions:
Traceback (most recent call last):
  File "/home/fr/.local/share/virtualenvs/sk/lib/python3.5/site-packages/sphinx/cmdline.py", line 243, in main
    opts.warningiserror, opts.tags, opts.verbosity, opts.jobs)
  File "/home/fr/.local/share/virtualenvs/sk/lib/python3.5/site-packages/sphinx/application.py", line 197, in __init__
    self._init_builder(self.buildername)
  File "/home/fr/.local/share/virtualenvs/sk/lib/python3.5/site-packages/sphinx/application.py", line 269, in _init_builder
    self.emit('builder-inited')
  File "/home/fr/.local/share/virtualenvs/sk/lib/python3.5/site-packages/sphinx/application.py", line 536, in emit
    results.append(callback(self, *args))
  File "/home/fr/.local/share/virtualenvs/sk/lib/python3.5/site-packages/sphinx/ext/autosummary/__init__.py", line 564, in process_generate_options
    base_path=app.srcdir)
  File "/home/fr/.local/share/virtualenvs/sk/lib/python3.5/site-packages/sphinx/ext/autosummary/generate.py", line 115, in generate_autosummary_docs
    items = find_autosummary_in_files(sources)
  File "/home/fr/.local/share/virtualenvs/sk/lib/python3.5/site-packages/sphinx/ext/autosummary/generate.py", line 225, in find_autosummary_in_files
    errors='ignore') as f:
  File "/home/fr/.local/share/virtualenvs/sk/lib/python3.5/codecs.py", line 895, in open
    file = builtins.open(filename, mode, buffering)
FileNotFoundError: [Errno 2] No such file or directory: '/home/fr/github/scikit-image/doc/source/auto_examples/transform/plot_radon_transform.txt.rst'
```

sciunto commented:
I start again, perhaps a make clean first will help.

emmanuelle commented:
@sciunto fingers crossed :-).

sciunto commented:
It works :)

It looks really cool. 

We will probably need to enhance examples at few places, like plot_gabors_from_astronaut. For some reason, the "preview" stops at the double quote for "sparse". We can use italic instead. 

Also, a lot of examples starts by "In this example, blabla" and it would be better to rephrase the first sentence. But we can do that separately.

There is also "IPython" printed in the download section that we can rename to avoid the confusion, especially for newcomers.

sciunto commented:
@emmanuelle Maybe we can fill up a new ticket to remember to remove files in scikit-image/doc/ext/sphinx_gallery once sphinx makes a new release.

soupault commented:
@sciunto I agree, many example have to be refactored after this. As for me, the intro's in the popups were quite messy.

@emmanuelle, thanks, that's very cool! Some comments:
- `features_detection/plot_holes_and_peaks.html` shows only the first plot. I haven't checked every single example, but there may be some extra cases like this;
- can we have `Script output` section as in, for example, http://scikit-learn.org/stable/auto_examples/calibration/plot_calibration.html ?

sciunto commented:
I have few questions, not necessarily directly related to this PR:

* Is it possible to have URL on function calls in examples to our API, like for scikit-learn? Ex: http://scikit-learn.org/stable/auto_examples/plot_cv_predict.html#example-plot-cv-predict-py
* Is it possible to choose which figure is used for the gallery? For instance, [here](http://scikit-image.org/docs/dev/auto_examples/features_detection/plot_holes_and_peaks.html#example-features-detection-plot-holes-and-peaks-py), it would be better to use the last one. If we can choose only globally, I have the feeling that the last one would be a better choice on average.

sciunto commented on doc/examples/features_detection/plot_holes_and_peaks.py:
```diff
@@ -72,7 +54,22 @@
 rec = reconstruction(seed, mask, method='dilation')
 
-ax[3].imshow(image-rec)
+fig,ax = plt.subplots(2, 2, figsize=(5, 4), sharex=True, sharey=True, subplot_kw={'adjustable':'box-forced'})
```
PEP8: fig, ax (space)
PEP8: line length limit?

sciunto commented on doc/examples/features_detection/plot_gabors_from_astronaut.py:
```diff
@@ -3,7 +3,7 @@
 ============================================================
 
-How to build a (bio-plausible) "sparse" dictionary (or 'codebook', or
+How to build a (bio-plausible) sparse dictionary (or 'codebook', or
```
sparce in italic?

soupault commented:
@emmanuelle sorry, did you have a chance to investigate on points mentioned by myself and @sciunto ?

I cannot wait to see this in `master` :rocket:. Moreover, many other PRs are ready and awaiting this awesome feature to be deployed :wink:.

emmanuelle commented:
@soupault me too :-).

About script output: it appears as an output cell of Ipython notebook, see for example ``plot_radon_transform.html``.

I've changed ``features_detection/plot_holes_and_peaks.html`` to improve the output.

@sciunto PEP 8 and small fixes done. With ``sphinx_gallery``, we do have now links to the API doc of functions of scikit-image, matplotlib, scipy... Is this what you were mentioning?

About choosing which figure is displayed as a thumbnail, it would be possible, but it should be submitted as a PR to ``sphinx_gallery``. I hope we can do it later, and that it does not block this PR.

So I hope I have fixed everything that was asked for!! :-)

sciunto commented:
OK I was tricked by the color used for links. I thought it was just highlighting and not a link.

I'm :+1:  to merge this. We can look at the image picked up in the summary in another issue.

soupault commented:
@emmanuelle great! Sorry, I somehow missed couple of your last commits.

Before merging this, does anyone of you have an idea how to manage old files (`*.txt`) removal from the website? I am a bit worried about the cached pages, because they appear higher in Google search results. See https://github.com/scikit-image/scikit-image/issues/2059 for the details.

sciunto commented:
Google will update himself as a big boy. I'm more worried by tutorials for ex. My server is regularly checking for broken links in the doc: https://linkchecker.sciunto.org/scikit-image.org-docs_stable https://linkchecker.sciunto.org/scikit-image.org-docs_dev I can do it for other websites as well.

soupault commented:
@sciunto oh, yes, this is a very good point! I guess me should consider adding a deprecation path for the website documentation.

By the way, are you sure your scanner works well? :wink: https://github.com/scikit-image/scikit-image/issues/1409

sciunto commented:
stable points to 0.12 not 0.11.

soupault commented on doc/examples/xx_applications/plot_morphology.py:
```diff
@@ -74,95 +71,82 @@ def plot_comparison(original, filtered, filter_name):
+#
+#Morphological ``dilation`` sets a pixel at (i, j) to the *maximum over all
+#pixels in the neighborhood centered at (i, j)*. Dilation enlarges bright
+#regions and shrinks dark regions.
```
Please, add a whitespace after `#` here and above.

soupault commented on doc/examples/xx_applications/plot_morphology.py:
```diff
@@ -74,95 +71,82 @@ def plot_comparison(original, filtered, filter_name):
+#
+#Morphological ``opening`` on an image is defined as an *erosion followed by
+#a dilation*. Opening can remove small bright spots (i.e. "salt") and
+#connect small dark cracks.
```
Same

soupault commented on doc/examples/xx_applications/plot_morphology.py:
```diff
@@ -74,95 +71,82 @@ def plot_comparison(original, filtered, filter_name):
+#connect small bright cracks.
+#
+#To illustrate this more clearly, let's add a small crack to the white
+#border:
```
Same

soupault commented on doc/examples/xx_applications/plot_morphology.py:
```diff
@@ -171,86 +155,70 @@ def plot_comparison(original, filtered, filter_name):
-
-"""
+######################################################################
+#As you can see, the 10-pixel wide black square is highlighted since
```
Also here and below :).

soupault commented on doc/source/conf.py:
```diff
@@ -31,11 +31,32 @@
               'numpydoc',
               'sphinx.ext.autosummary',
-              'plot2rst',
+              #'plot2rst',
```
This should be removed?

soupault commented on doc/examples/color_exposure/plot_ihc_color_separation.py:
```diff
@@ -62,14 +60,11 @@
 zdh = np.dstack((np.zeros_like(h), d, h))
 
-#fig, ax = plt.subplots()
+# fig, ax = plt.subplots()
```
This as well :).

soupault commented:
@sciunto of course, sorry. Worth trying to run your scanner on 0.11 then ( IMO :) ).

@emmanuelle sorry for attacking your patience :smiley:. I've added just a couple of concluding remarks.

emmanuelle commented:
@soupault no worries, thank you for the additional review. I removed the commented lines. About the missing spaces, it's some sphinx / sphinx-gallery magic: the layout is broken if I add the spaces. Maybe someone else can figure out if this can be improved. At the moment I'm in the middle of a 24/7 synchrotron campaign, so catching up with PRs is not easy. I let you guys decide if it can be merged as such.

soupault commented:
@emmanuelle this builds and works fine for me https://gist.github.com/soupault/170e3065384431ae438e3793ac185fc2 . You have whitespaces after `#` for any other example, haven't you?

emmanuelle commented:
If you can improve this, it's perfect, but I have no time to fix it in the next few days (experiments running 24/7). Could you please submit a pull request to my branch, or if the PR can be merged with this minor style problem, file a separate PR later on? Thanks!

soupault commented:
@emmanuelle I understood. Examples will need an extra review anyways, for me it is a no blocker.

I think the PR is finalized in it's essential part. As we have :+1: from @sciunto and several new PRs are starting to depend on this, I'm going for a merge.

@emmanuelle thank you a lot for this awesome work, and sorry again for any offense made! :wine_glass:

soupault commented:
@emmanuelle could you squash the commits, please? It is not possible to do this via GitHub interface ('Not enabled for this repository' :confused: )

emmanuelle commented:
@soupault I squashed most commits together, I only left the most important ones at the beginning because it may be important to keep a bit of granularity for an important move.

soupault commented:
@emmanuelle Build fails on OS X.
Maybe worth squashing also https://github.com/scikit-image/scikit-image/pull/2078/commits/94d051c124b03434e2c999ac5c674149f1f31fcd and https://github.com/scikit-image/scikit-image/pull/2078/commits/d998e7e5341b4c8f537ade6c6a66767d8c763846 as the former is performing files' renaming while the latter is undoing it.

sciunto commented:
It's a latex problem

```

[31;01mWarning, treated as error:[39;49;00m

WARNING: display latex '\\sqrt{2}sigma': latex exited with error

[stdout]

This is pdfTeX, Version 3.14159265-2.6-1.40.17 (TeX Live 2016) (preloaded format=latex)

 restricted \write18 enabled.

entering extended mode

(./math.tex

LaTeX2e <2016/03/31>

Babel <3.9r> and hyphenation patterns for 22 language(s) loaded.

(/usr/local/texlive/2016basic/texmf-dist/tex/latex/base/article.cls

Document Class: article 2014/09/29 v1.4h Standard LaTeX document class

(/usr/local/texlive/2016basic/texmf-dist/tex/latex/base/size12.clo))

(/usr/local/texlive/2016basic/texmf-dist/tex/latex/base/inputenc.sty



! LaTeX Error: File `utf8x.def' not found.



Type X to quit or <RETURN> to proceed,

or enter new name. (Default extension: def)



Enter file name: 

! Emergency stop.

<read *> 

         

l.158 \endinput

               ^^M

No pages of output.

Transcript written on math.log.
```

sciunto commented:
@soupault Looks like it's travis. Same problem here #2129

soupault commented:
This is passing now.
Builds on OS X started to fail 2 days ago, not sure that is the cause... Let's give a life to this PR and then investigate that issue.

emmanuelle commented:
Thank you !!!

stefanv commented:
This was a massive amount of work; thank you so much @emmanuelle and @soupault for diligent reviewing.

stefanv commented:
Quick question: is this kind of output expected?

```
plotting code blocks in ../examples/transform/plot_seam_carving.py
 - time elapsed : 0.12 sec
plotting code blocks in ../examples/transform/plot_seam_carving.py
 - time elapsed : 0.12 sec
plotting code blocks in ../examples/transform/plot_seam_carving.py
 - time elapsed : 1.2 sec
plotting code blocks in ../examples/transform/plot_seam_carving.py
 - time elapsed : 0.11 sec
plotting code blocks in ../examples/transform/plot_seam_carving.py
```

emmanuelle commented:
Yes, this output comes from sphinx_gallery. It can be useful to detect examples that take a lot of time to build. If you think that this is too verbose, we could submit a PR to sphinx_gallery so that this output can be optional only.

emmanuelle commented:
@soupault do you mean that building the doc (make html) is failing for you on OS X?

soupault commented:
@emmanuelle no, not for me. There were some issues with our OS X Travis build (actually, I think that was Travis bug). See https://github.com/scikit-image/scikit-image/pull/2078#issuecomment-224073869 for the error message.
